### PR TITLE
Add SABER

### LINKED ISSUE #114: Add SABER

### PR DESCRIPTION
**Pull request by taudor**
Add Saber KEM.

pqm4/SUPERCOP benchmarks not included, only optimized implementation from the round 2 submission paper.

Closes #114

### CODE REVIEW

File: encryption/saber/saberKEM-sha3/bench/avx2_FireSABER_skylake.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+platform: 'Intel Core i7-6600U (Skylake) CPU 2.60GHz, Ubuntu, gcc v. 7.0'
```
Comment (dingens) [inline encryption/saber/saberKEM-sha3/bench/avx2_FireSABER_skylake.yaml:1]: They are contradicting themselves. In the text, they say i7-6600, in the Table caption, they say i5-7200. What do we do about this?

File: encryption/saber/saberKEM-sha3/param/SABER.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+name: SABER
+security level:
+  classical: 203
+  quantum: 185
+  nist category: 3
+  comment: 'See https://www.esat.kuleuven.be/cosic/pqcrypto/saber/index.html'
+failure probability: -136
+sizes:
+  sk: 2304
```
Comment (dingens) [inline encryption/saber/saberKEM-sha3/param/SABER.yaml:9]: shall we add the compressed size as a comment?

File: encryption/saber/saberKEM-sha3/bench/avx2_FireSABER_skylake.yaml (hunk 1 above)
Comment (taudor) [reply encryption/saber/saberKEM-sha3/bench/avx2_FireSABER_skylake.yaml:1]: I compared the round 1 and round 2 submissions: the table caption *does* change, but the text paragraph *does not* from v1 to v2.
So I updated the information to match the table caption from v2.

File: encryption/saber/saberKEM-sha3/param/SABER.yaml (hunk 2 above)
Comment (taudor) [reply encryption/saber/saberKEM-sha3/param/SABER.yaml:9]: Added the information

**Comment by taudor** [comment]
@dingens I added missing/additional information.

I don't see a comparison use case for the PKE. We are already missing some of the PKEs from other schemes. For SABER we would only have the parameter sets to add without explicit benchmarks.

File: encryption/saber/saberKEM-sha3/bench/avx2_FireSABER_skylake.yaml (hunk 1 above)
Comment (dingens) [reply encryption/saber/saberKEM-sha3/bench/avx2_FireSABER_skylake.yaml]: smart move :)

File: encryption/saber/saberKEM-sha3/param/LightSABER.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+name: LightSABER
+security level:
+  classical: 125
+  quantum: 114
+  nist category: 1
+  comment: 'See https://www.esat.kuleuven.be/cosic/pqcrypto/saber/index.html'
```
Comment (dingens) [inline encryption/saber/saberKEM-sha3/param/LightSABER.yaml:6]: I don't understand why you added this extra note?

Comment (taudor) [reply encryption/saber/saberKEM-sha3/param/LightSABER.yaml:6]: So, I should've added more information, since this is kind of confusing and I don't know if we should add additional third party information.

While adding SABER I compared the information from the homepage with the submission paper. And the security numbers are off. 
So [here](https://estimate-all-the-lwe-ntru-schemes.github.io/docs/) are security estimations about LWE/NTRUE schemes. And these numbers are shown on the SABER homepage (but not all correctly). These are other assumption numbers than in the submission paper. This is also true for other schemes that we already included, but the numbers still don't match all the time (even the claimed security).

Long story short I think I should add the security bits from the original submission paper and delete the other values. This is also consistent with the other schemes.

Comment (dingens) [reply encryption/saber/saberKEM-sha3/param/LightSABER.yaml:6]: About those comic sans ms ;) values: agreed that we should not add those if they differ.

About the paper I linked: I'd say it's not really external (two common authors) so I'd say we can and should include it.

Comment (taudor) [reply encryption/saber/saberKEM-sha3/param/LightSABER.yaml:6]: I agree that we should include the paper that you linked. However, I'm not sure how useful the data is for comparison. The paper includes only timings for the decapsulation function for one parameter set. So we are missing key generation and encapsulation, and two parameter sets.
They're writing that the benchmarks are included in pqm4, but I couldn't find them.

For now, I would include SABER without additional data. We can have a separate look, or the data still shows up in pqm4.

Comment (dingens) [reply encryption/saber/saberKEM-sha3/param/LightSABER.yaml]: > The paper includes only timings for the decapsulation function for one parameter set

ah, ok, I didn't look into it that deep. Agreed, then.

PQM4 is a separate issue anyways. I'll re-run the imports next week or so when we merged everything.

**Comment by dingens** [comment]
I don't get where these numbers in 853d69c are from? Both the website and the pdf mention 283/257 (for firesaber), not the 245/270 you replaced it with. I don't find the latter numbers anywhere. (Ah, they seem to be the round 1 numbers: https://www.google.de/search?tbs=li:1&q=saber%20270%20245%20126%20115%20198%20180)

**Comment by taudor** [comment]
> I don't get where these numbers in [853d69c](https://github.com/cryptoeng/pqdb/commit/853d69c89f702f4c3f02017c926f1b2f5e518e61) are from? Both the website and the pdf mention 283/257 (for firesaber), not the 245/270 you replaced it with. I don't find the latter numbers anywhere. (Ah, they seem to be the round 1 numbers: https://www.google.de/search?tbs=li:1&q=saber%20270%20245%20126%20115%20198%20180)

Sorry, got confused with the multiple pdf files while fixing the other problem. I got the round 1 numbers by accident. Now it should be fine.
